src/Microsoft.Extensions.WebEncoders.Core/UrlPathDecoder.cs

```diff
@@ -0,0 +1,341 @@
+        {
+            var bufferPosition = 0;
+            var sourcePosition = 0;
+
+            var unescapedChars = new char[1];
+            var unescapedCharsCount = 0;
+            var bytesBuffer = new byte[4];
+
+            while (sourcePosition < source.Length)
+            {
+                var next = GetNextEncoded(sourcePosition, source);
+                var copyLength = next - sourcePosition;
+                Copy(source, ref sourcePosition, destination, ref bufferPosition, copyLength);
+
+                if (sourcePosition >= source.Length)
+                {
+                    break;
+                }
+
+                var consumed = Unescape(source, next, bytesBuffer, ref unescapedChars, ref unescapedCharsCount);
+                if (consumed == 0)
+                {
+                    // Skip unescaping the % as the sequence follows it can't be correctly
+                    // decoded under UTF8
+                    Copy(source, ref sourcePosition, destination, ref bufferPosition, 1);
+                }
+                else if (unescapedCharsCount == 1 && SkipUnescape(unescapedChars[0]))
+                {
+                    // Skip unescaping specified characters (eg. '/')
+                    // Copy the original sequence to destination
+                    Copy(source, ref sourcePosition, destination, ref bufferPosition, consumed);
+                }
+                else
+                {
+                    // Copy unescaped chararter. Move to the next charactor in source.
+                    for (int i = 0; i < unescapedCharsCount; ++i)
+                    {
+                        destination[bufferPosition++] = unescapedChars[i];
+                    }
+                    sourcePosition += consumed;
+                }
+            }
+
+            return bufferPosition;
+        }
+
+        private static bool SkipUnescape(char charactor)
+        {
+            if (charactor == '/')
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Unescape a sequence of characters
+        /// The sequence is a substring of the given source. The start index must point to
+        /// a % character initializes the sequence.
+        /// If the sequence following the % can be successfully decoded in UTF8,
+        /// - The result char will be set to the unescaped character
+        /// - The length of the sequence, including the % charactor, will be returned.
+        /// Otherwise 0 is returned.
+        /// </summary>
+        private static int Unescape(char[] source, int start, byte[] bytesBuffer, ref char[] output, ref int count)
+        {
+            if (start + 2 >= source.Length)
+            {
+                return 0;
+            }
+
+            byte first;
+            if (!TryGetUnescapedByte(source[start + 1], source[start + 2], out first))
```

Comment:
This is always called with `source[start + 1], source[start + 2]`. Pass in the array and offset instead and do the Length check internally. I see a few places where you may have missed the length check.